File: packages/business-logic/src/money.ts
/**
 * Money utilities.
 * All internal values are integer cents (e.g. $12.34 = 1234).
 * Use Decimal.js for any arithmetic to avoid floating-point errors.
 */
import Decimal from "decimal.js";

Decimal.set({ precision: 20, rounding: Decimal.ROUND_HALF_UP });

/** Add two cent values safely */
export function addCents(a: number, b: number): number {
  return new Decimal(a).plus(new Decimal(b)).toNumber();
}

/** Subtract cent values safely */
export function subtractCents(a: number, b: number): number {
  return new Decimal(a).minus(new Decimal(b)).toNumber();
}

/** Multiply cents by a rate (e.g. for interest calculations) */
export function multiplyCents(cents: number, rate: number): number {
  return new Decimal(cents).times(new Decimal(rate)).toDecimalPlaces(0).toNumber();
}

/** Format cents as a locale currency string: 1234 → "$12.34" */
export function formatCurrency(
  cents: number,
  currency = "USD",
  locale = "en-US"
): string {
  return new Intl.NumberFormat(locale, {
    style: "currency",
    currency,
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(cents / 100);
}

/** Format cents as a plain decimal string: 1234 → "12.34" */
export function centsToString(cents: number): string {
  return (cents / 100).toFixed(2);
}

/** Parse a user-entered string to cents: "12.34" → 1234 */
export function parseToCents(value: string): number {
  const cleaned = value.replace(/[^0-9.-]/g, "");
  return new Decimal(cleaned).times(100).toDecimalPlaces(0).toNumber();
}

/** Calculate credit utilization as a percentage (0–100) */
export function creditUtilization(balance: number, limit: number): number {
  if (limit === 0) return 0;
  return new Decimal(Math.abs(balance))
    .div(new Decimal(limit))
    .times(100)
    .toDecimalPlaces(1)
    .toNumber();
}

/** Calculate simple monthly interest in cents */
export function monthlyInterest(balanceCents: number, aprPercent: number): number {
  const monthlyRate = new Decimal(aprPercent).div(100).div(12);
  return new Decimal(Math.abs(balanceCents))
    .times(monthlyRate)
    .toDecimalPlaces(0)
    .toNumber();
}

File: mobile/lib/core/utils/money.dart
// Money utility functions.
//
// All monetary values in the database are stored as INTEGER CENTS to avoid
// floating-point rounding errors (e.g. $12.34 is stored as 1234).
// These helpers convert between cents and display-ready strings.
import 'package:decimal/decimal.dart';
import 'package:intl/intl.dart';

/// Formats integer cents as a locale-aware currency string.
/// Example: 1234 → "$12.34"
String formatCurrency(int cents, {String currency = 'USD', String locale = 'en_US'}) {
  final formatter = NumberFormat.currency(locale: locale, symbol: r'$', decimalDigits: 2);
  return formatter.format(cents / 100);
}

/// Formats cents as a plain decimal string without a currency symbol.
/// Uses [Decimal] to avoid floating-point errors during division.
/// Example: 1234 → "12.34"
String centsToString(int cents) {
  return (Decimal.fromInt(cents) / Decimal.fromInt(100))
      .toDecimal(scaleOnInfinitePrecision: 2)
      .toStringAsFixed(2);
}

/// Parses a user-entered string (e.g. "$12.34" or "12.34") to integer cents.
/// Strips non-numeric characters before parsing so currency symbols are safe.
/// Example: "12.34" → 1234
int parseToCents(String value) {
  final cleaned = value.replaceAll(RegExp(r'[^\d.]'), '');
  if (cleaned.isEmpty) return 0;
  // Multiply by 100 using Decimal to avoid floating-point rounding, then
  // convert to BigInt before toInt() because Decimal doesn't expose toInt().
  final d = Decimal.parse(cleaned) * Decimal.fromInt(100);
  return d.round().toBigInt().toInt();
}

/// Returns credit utilization as a percentage (0–100+).
/// A result over 100 means the card is over its limit.
double creditUtilization(int balanceCents, int limitCents) {
  if (limitCents == 0) return 0;
  return (balanceCents.abs() / limitCents * 100).clamp(0, 999).toDouble();
}

/// Returns true if [cents] represents a debit (money leaving an account).
bool isDebit(int cents) => cents < 0;

/// Returns true if [cents] is zero or positive (income or zero balance).
bool isPositive(int cents) => cents >= 0;
